File: src/xunit.installer.templates/MVC/VisualBasic/MVC1-VS2008/Routes/RouteFacts.vb
Imports System.Web
Imports System.Web.Routing
Imports Xunit

Public Class RouteFacts
    <Fact()> Public Sub RouteWithControllerNoActionNoId()
        ' Arrange
        Dim context = New StubContext("~/controller1")
        Dim routes = New RouteCollection()
        MvcApplication.RegisterRoutes(routes)

        ' Act
        Dim routeData = routes.GetRouteData(context)

        ' Assert
        Assert.NotNull(routeData)
        Assert.Equal("controller1", routeData.Values("controller"))
        Assert.Equal("Index", routeData.Values("action"))
        Assert.Equal("", routeData.Values("id"))
    End Sub

    <Fact()> Public Sub RouteWithControllerWithActionNoId()
        ' Arrange
        Dim context = New StubContext("~/controller1/action2")
        Dim routes = New RouteCollection()
        MvcApplication.RegisterRoutes(routes)

        ' Act
        Dim routeData = routes.GetRouteData(context)

        ' Assert
        Assert.NotNull(routeData)
        Assert.Equal("controller1", routeData.Values("controller"))
        Assert.Equal("action2", routeData.Values("action"))
        Assert.Equal("", routeData.Values("id"))
    End Sub

    <Fact()> Public Sub RouteWithControllerWithActionWithId()
        ' Arrange
        Dim context = New StubContext("~/controller1/action2/id3")
        Dim routes = New RouteCollection()
        MvcApplication.RegisterRoutes(routes)

        ' Act
        Dim routeData = routes.GetRouteData(context)

        ' Assert
        Assert.NotNull(routeData)
        Assert.Equal("controller1", routeData.Values("controller"))
        Assert.Equal("action2", routeData.Values("action"))
        Assert.Equal("id3", routeData.Values("id"))
    End Sub

    <Fact()> Public Sub RouteWithTooManySegments()
        ' Arrange
        Dim context = New StubContext("~/a/b/c/d")
        Dim routes = New RouteCollection()
        MvcApplication.RegisterRoutes(routes)

        ' Act
        Dim routeData = routes.GetRouteData(context)

        ' Assert
        Assert.Null(routeData)
    End Sub

    <Fact()> Public Sub RouteForEmbeddedResource()
        ' Arrange
        Dim context = New StubContext("~/foo.axd/bar/baz/biff")
        Dim routes = New RouteCollection()
        MvcApplication.RegisterRoutes(routes)

        ' Act
        Dim routeData = routes.GetRouteData(context)

        ' Assert
        Assert.NotNull(routeData)
        Assert.IsAssignableFrom(Of StopRoutingHandler)(routeData.RouteHandler)
    End Sub
End Class

File: src/xunit.installer.templates/MVC/CSharp/MVC3-VS2010/Routes/RouteFacts.cs
using System.Web;
using System.Web.Mvc;
using System.Web.Routing;
using Xunit;
using $mvcprojectnamespace$;

namespace $safeprojectname$.Routes
{
    public class RouteFacts
    {
        [Fact]
        public void RouteWithControllerNoActionNoId()
        {
            // Arrange
            StubContext context = new StubContext("~/controller1");
            RouteCollection routes = new RouteCollection();
            MvcApplication.RegisterRoutes(routes);

            // Act
            RouteData routeData = routes.GetRouteData(context);

            // Assert
            Assert.NotNull(routeData);
            Assert.Equal("controller1", routeData.Values["controller"]);
            Assert.Equal("Index", routeData.Values["action"]);
            Assert.Equal(UrlParameter.Optional, routeData.Values["id"]);
        }

        [Fact]
        public void RouteWithControllerWithActionNoId()
        {
            // Arrange
            StubContext context = new StubContext("~/controller1/action2");
            RouteCollection routes = new RouteCollection();
            MvcApplication.RegisterRoutes(routes);

            // Act
            RouteData routeData = routes.GetRouteData(context);

            // Assert
            Assert.NotNull(routeData);
            Assert.Equal("controller1", routeData.Values["controller"]);
            Assert.Equal("action2", routeData.Values["action"]);
            Assert.Equal(UrlParameter.Optional, routeData.Values["id"]);
        }

        [Fact]
        public void RouteWithControllerWithActionWithId()
        {
            // Arrange
            StubContext context = new StubContext("~/controller1/action2/id3");
            RouteCollection routes = new RouteCollection();
            MvcApplication.RegisterRoutes(routes);

            // Act
            RouteData routeData = routes.GetRouteData(context);

            // Assert
            Assert.NotNull(routeData);
            Assert.Equal("controller1", routeData.Values["controller"]);
            Assert.Equal("action2", routeData.Values["action"]);
            Assert.Equal("id3", routeData.Values["id"]);
        }

        [Fact]
        public void RouteWithTooManySegments()
        {
            // Arrange
            StubContext context = new StubContext("~/a/b/c/d");
            RouteCollection routes = new RouteCollection();
            MvcApplication.RegisterRoutes(routes);

            // Act
            RouteData routeData = routes.GetRouteData(context);

            // Assert
            Assert.Null(routeData);
        }

        [Fact]
        public void RouteForEmbeddedResource()
        {
            // Arrange
            StubContext context = new StubContext("~/foo.axd/bar/baz/biff");
            RouteCollection routes = new RouteCollection();
            MvcApplication.RegisterRoutes(routes);

            // Act
            RouteData routeData = routes.GetRouteData(context);

            // Assert
            Assert.NotNull(routeData);
            Assert.IsAssignableFrom<StopRoutingHandler>(routeData.RouteHandler);
        }
    }
}